OSOE-400: Auto-generate next version number when publishing NuGet packages

AydinE opened the pull request:
[OSOE-400](https://lombiq.atlassian.net/browse/OSOE-400)

[OSOE-400]: https://lombiq.atlassian.net/browse/OSOE-400?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AydinE committed:
Automatic version bump scripts

AydinE committed:
Fix Powershell analyzer errors

AydinE committed:
Update VNext.ps1

AydinE committed:
Spelling

AydinE committed:
Update VNext.ps1

AydinE committed:
Add interactive mode and readme

AydinE committed:
Update VNext.ps1

AydinE committed:
Update Releasing.md

sarahelsaig commented on Get-Solution-Projects.ps1:
Please rename this file to _Get-SolutionProject.ps1_ to follow the Powershell cmdlet naming conventions
([the noun part should be a single PascalCase word](https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/strongly-encouraged-development-guidelines?view=powershell-7.4#use-pascal-case-for-cmdlet-names-sd02) and [it should be singular instead of plural](https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/strongly-encouraged-development-guidelines?view=powershell-7.4#use-singular-parameter-names)).

sarahelsaig commented on Get-Solution-Projects.ps1:
```diff
@@ -0,0 +1,42 @@
+if (-not $NewVersion)
+{
+    throw 'No new version specified.'
+}
```
Mark the parameter mandatory instead of manually checking with `[Parameter(Mandatory=$true)]`([see docs](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_advanced_parameters?view=powershell-7.4#parameter-attribute))

sarahelsaig commented on Get-Solution-Projects.ps1:
```diff
@@ -0,0 +1,42 @@
+        # Add the project name to the array
+        $projectNames += $projectName
+    }
+}
```
This is not necessary, you can use the dotnet CLI like `dotnet sln <SLN_FILE> list`.

sarahelsaig commented on Update-References.ps1:
```diff
@@ -0,0 +1,39 @@
+param(
+    [string]$ProjectToFind, # Name of the project to check reference for
+    [string]$NewVersion # New version number to set
+)
```
Use [comment based help](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_comment_based_help?view=powershell-7.4) instead of regular comments. It's more structured and `Get-Help` will display it as well.

sarahelsaig commented on Update-References.ps1:
```diff
@@ -0,0 +1,39 @@
+$projectFiles = Get-ChildItem -Path .\ -Recurse -Filter '*.csproj'
+
+# Loop through each project file to see if it has a reference to the project we are looking for
+ForEach ($projectFile in $projectFiles)
```
You use `foreach` all lower case everywhere else ([and so does the documentation](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_foreach?view=powershell-7.4)). Let's keep it consistent.

```suggestion
foreach ($projectFile in $projectFiles)
```

sarahelsaig commented on Update-References.ps1:
```diff
@@ -0,0 +1,39 @@
+)
+
+# Get all .csproj files recursively
+$projectFiles = Get-ChildItem -Path .\ -Recurse -Filter '*.csproj'
```
Path is assumed to be `.` by default so that's not necessary.

```suggestion
$projectFiles = Get-ChildItem -Recurse -Filter '*.csproj'
```

sarahelsaig commented on Update-References.ps1:
```diff
@@ -0,0 +1,39 @@
+    # Find the ProjectReference nodes
+    $projectReferences = $projectXml.Project.ItemGroup.PackageReference
+
+    foreach ($projectReference in $projectReferences)
```
Since this is for `<PackageReference>` rename these to `$packageReference` and `$packageReferences` respectively.

sarahelsaig commented on Update-References.ps1:
```diff
@@ -0,0 +1,39 @@
+        # Check if the Include attribute contains the project name
+        if ($projectReference.Include -like "*$ProjectToFind*")
+        {
+            if ($projectReference.Version)
```
Merge these two nested conditions with `-and`.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+# Call like this: .\VNext.ps1 -Path '\Path\to\project' -PreRelease $True -Issue '1234' -UpdateReferences $true
```
You use both `$True` and `$true` across the PR. Please replace all `$True` with `$true` for consistency.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+# Call like this: .\VNext.ps1 -Path '\Path\to\project' -PreRelease $True -Issue '1234' -UpdateReferences $true
+# or like this: .\VNext.ps1 -Path '\Path\to\project' -Type 'major' -UpdateReferences $true
```
Replace these with comment based help as well.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+$origin = Get-Location
+$newVersion = ''
+
+if ($PreRelease -eq $false -and -not $PSBoundParameters.ContainsKey('Type'))
```
Is there a reason why you used `$PreRelease -eq $false` instead of `-not $PreRelease`?

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+    [string]$Issue,
+    [bool]$PreRelease = $false,
+    [bool]$UpdateReferences = $false,
+    [bool]$Interactive = $true
```
Please turn these into [`[switch]` parameters](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_functions_advanced_parameters?view=powershell-7.4#switch-parameters) and replace `[bool]$Interactive = $true` with `[switch]$NonInteractive` so that can be false by default as well. It's best to use switch instead of bool, because then you can invoke it like
```pwsh
.\VNext.ps1 -Path '\Path\to\project' -Type 'major' -UpdateReferences
```
without having to pass the `$true`. It's how the built-in cmdlets behave so it's more familiar to the user.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+    throw 'Must set a version Type (major, minor, patch)'
+    if ($Type -ne 'major' -and $Type -ne 'minor' -and $Type -ne 'patch')
+    {
+        throw 'Invalid version Type must be major, minor, or patch'
```
Put the values in quotes.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+    }
+}
+
+if ($PreRelease -eq $true -and -not $PSBoundParameters.ContainsKey('Issue'))
```
```suggestion
if ($PreRelease -and -not $PSBoundParameters.ContainsKey('Issue'))
```

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+}
+
+#Get the latest release tags, we get the last 20 tags so we can determine the next version or pre-release version
+$latestTags = git for-each-ref --sort=-creatordate --count=20 --format '%(refname:short)' refs/tags
```
Why is the number 20 necessary here? It works for me without it just as well. Won't this break if there were more than  20 prerelease versions?

sarahelsaig replied:
@AydinE you didn't respond to this.

AydinE replied:
@sarahelsaig Sorry must have missed this one, I only have 20 as an arbitrary number really. This could have been any number and is really only in some case where maybe a (pre-)release of another issue was published in between 2 versions of the current issue. I am not sure how common that would be or how much of an edge case that is but I built it like that just in case. 

So in case we were both working on issues in the same repo and you released v1.1.1-alpha.1.osoe-123 and I published v1.1.1-alpha.1.osoe-321 and then you wanted to do v1.1.1-alpha.2.osoe-123 after that it would not work without this. But perhaps this is overthinking it?

sarahelsaig replied:
> Sorry must have missed this one

If you put an emoji reaction (like :+1: or :eyes:) on a comment you've already addressed then it's easy to spot if there is anything left out.

> This could have been any number and is really only in some case where maybe a (pre-)release of another issue was published in between 2 versions of the current issue.

But why constrain it at all? If you remove the `--count=20` argument it still works and we will never have so many tags where fetching the compete list could cause an issue.

> So in case we were both working on issues in the same repo and you released v1.1.1-alpha.1.[osoe-123](https://lombiq.atlassian.net/browse/OSOE-123) and I published v1.1.1-alpha.1.[osoe-321](https://lombiq.atlassian.net/browse/OSOE-321) and then you wanted to do v1.1.1-alpha.2.[osoe-123](https://lombiq.atlassian.net/browse/OSOE-123) after that it would not work without this. But perhaps this is overthinking it?

You have `--sort=-creatordate` and filter the results by `$Issue` so I don't think that can ever be a problem.

AydinE replied:
Good call, I have removed the arbitrary count constraint now

sarahelsaig committed:
Fix comment spacing.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+# Check tags for a valid version number and try to determine the next version
+try
+{
+    if ($latestTags[0] -match 'v\d+\.\d+\.\d+')
```
Check from string start.
```suggestion
    if ($latestTags[0] -match '^v\d+\.\d+\.\d+')
```

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+# Check tags for a valid version number and try to determine the next version
+try
+{
+    if ($latestTags[0] -match 'v\d+\.\d+\.\d+')
```
Aren't you missing a foreach here? It seems to me that now it's only checking the first tag.

AydinE replied:
For this section I am indeed only checking the first tag, it's only needed to look at the rest of the tags in case of a pre-release (just in case there were other releases of the same package in between pre-releases for the same issue).
But let me know if I am missing some edge cases here maybe

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+        $version = $latestVersion -split '\.'
+        $major = [int]$version[0].Substring(1)
+        $minor = [int]$version[1]
+        $patch = [int]$version[2].Substring(0, 1)
```
It's safer to convert it into a `Version` instance with `[Version]$matches[0].Substring(1)`.

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+
+        $newVersion = "v$major.$minor.$patch"
+
+        if ($PreRelease -eq $true)
```
```suggestion
        if ($PreRelease)
```

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+        if ($PreRelease -eq $true)
+        {
+            # Check if there were any pre-release tags for this Issue already in latestTags array
+            $preReleaseTags = $latestTags | Where-Object { $PSItem -match ([regex]::escape($Issue)) }
```
```suggestion
            $preReleaseTags = $latestTags | Where-Object { $PSItem -match ([regex]::Escape($Issue)) }
```

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+            $preReleaseTags = $latestTags | Where-Object { $PSItem -match ([regex]::escape($Issue)) }
+            if ($preReleaseTags.Count -gt 0)
+            {
+                $latestPreRelease = $preReleaseTags[0]
```
If there was only one prerelease for an issue then `$preReleaseTags` will be `[string]` and `$latestPreRelease` will be just `"v"`:
![image](https://github.com/Lombiq/Open-Source-Orchard-Core-Extensions/assets/4007293/943f3b97-fde9-4bfb-89cf-be2253b5c775)

sarahelsaig commented on VNext.ps1:
```diff
@@ -0,0 +1,124 @@
+    }
+    else
+    {
+        Write-Output 'No valid tags found'
```
```suggestion
        Write-Output 'No valid tags were found.'
```

sarahelsaig commented on Releasing.md:
```diff
@@ -0,0 +1,90 @@
+﻿# Releasing a new version
+
+For releasing a new version of any project in this repository to NuGet there are a few handy scripts that you can use.
```
```suggestion
We have a few handy scripts you can use to release a new NuGet package version of any project in this repository.
```

sarahelsaig commented on Releasing.md:
````diff
@@ -0,0 +1,90 @@
+
+### Releasing a new major version
+
+```shell
````
`shell` refers to Unix _.sh_ scripts. Please replace with `pwsh` across the document.

sarahelsaig commented on Releasing.md:
````diff
@@ -0,0 +1,90 @@
+3. Run the script with the necessary parameters. For example to release a new major version of the `Lombiq.HelpfulLibraries` project and update the references in other projects, run the following command
+
+    ```shell
+    .\VNext.ps1 -path .\src\Libraries\Lombiq.HelpfulLibraries\ -type "major" -commitChanges $true
````
Make sure all the parameter names have the correct casing, e.g. `-Path`, `-Type`, etc. (same below)

sarahelsaig commented on VNext.ps1:
Consider renaming this file to `Publish-VNext` to match the code style of using [an approved verb](https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-7.4#data-verbs)

AydinE committed:
Cleanup powershell scripts and readme

AydinE committed:
style

AydinE committed:
Further changes

AydinE committed:
Update Publish-VNext.ps1

AydinE committed:
Handle casting to array

AydinE committed:
Update Publish-VNext.ps1

AydinE commented:
I believe I have addressed all of the points, since implementing your suggestion of using `dotnet sln list` which is obviously much cleaner I have decided the `Get-Solution-Projects` script was no longer needed so this is now removed.

AydinE committed:
Update Releasing.md

AydinE committed:
Remove arbitrary 20 count
